src/Tools/Source/RunTests/AssemblyScheduler.cs

```diff
@@ -2,359 +2,314 @@
+            // Now for our current set of test methods we got from the assemblies we built, match them to tests from our test run history
+            // so that we can extract an estimate of the test execution time for each test.
+            orderedTypeInfos = UpdateTestsWithExecutionTimes(orderedTypeInfos, testHistory);
+
+            // Create work items by partitioning tests by historical execution time with the goal of running under our time limit.
+            // While we do our best to run tests from the same assembly together (by building work items in assembly order) it is expected
+            // that some work items will run tests from multiple assemblies due to large variances in test execution time.
+            var workItems = BuildWorkItems<TimeSpan>(
+                orderedTypeInfos,
+                isOverLimitFunc: (accumulatedExecutionTime) => accumulatedExecutionTime >= s_maxExecutionTime,
+                addFunc: (currentTest, accumulatedExecutionTime) => currentTest.ExecutionTime + accumulatedExecutionTime);
+            LogWorkItems(workItems);
+            return workItems;
+
+            static ImmutableArray<WorkItemInfo> CreateWorkItemsForFullAssemblies(ImmutableArray<AssemblyInfo> assemblies)
+            {
+                var workItems = new List<WorkItemInfo>();
+                var partitionIndex = 0;
+                foreach (var assembly in assemblies)
                 {
-                    currentId++;
-                    currentTypeInfoList.Clear();
-                    currentClassNameLengthSum = 0;
-
-                    // Ensure the EventListenerGuard is in every partition.
-                    if (hasEventListenerGuard)
-                    {
-                        currentClassNameLengthSum += EventListenerGuardFullName.Length;
-                    }
+                    var currentWorkItem = ImmutableSortedDictionary<AssemblyInfo, ImmutableArray<TestMethodInfo>>.Empty.Add(assembly, ImmutableArray<TestMethodInfo>.Empty);
+                    workItems.Add(new WorkItemInfo(currentWorkItem, partitionIndex++));
                 }
 
-                void CheckForPartitionLimit(bool done)
-                {
-                    if (done)
-                    {
-                        // The builder is done looking at types.  If there are any TypeInfo that have not
-                        // been added to a partition then do it now.
-                        if (currentTypeInfoList.Count > 0)
-                        {
-                            FinishPartition();
-                        }
-
-                        return;
-                    }
-
-                    // One item we have to consider here is the maximum command line length in 
-                    // Windows which is 32767 characters (XP is smaller but don't care).  Once
-                    // we get close then create a partition and move on. 
-                    if (currentTypeInfoList.Sum(x => x.MethodCount) >= methodLimit ||
```

Comment:
Did this logic get added back somewhere? It's important. You may not hit the command line limit the first time you do this but the system is dynamic. The partitions will change as the code changes and if the stars align you will end up exceeding this limit. 

Found this out the hard way which is why this code exists.

Reply:
We are using rsp files to side step the command line length limit.

Reply:
this no longer applies :)  I run tests via [respone files](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-response-files?view=vs-2022) which don't really have a limit (other than bugs in the test platform like https://github.com/microsoft/vstest/issues/3905

Reply:
Ah didn't realize it supported it. When i first wrote this code it was against xunit console runner which didn't. Missed that vstest supported it when we flipped over to `dotnet test` based execution

Reply:
it generally works quite well, but does have a couple caveats - for example we can't just pass an RSP file containing our assemblies and filters to `dotnet test` - it doesn't properly forward the arguments to vstest.console.  So we instead directly invoke `vstest.console.dll` and pass it the RSP file.

See also this code - https://github.com/dotnet/roslyn/pull/62138/files#diff-f1dee33d3e9f22c56de807a8ebca67ca769712cc6b64d1e36c5a51d69f74b250R218